Source/Particles/Resampling/VelocityCoincidenceThinning.cpp

```diff
@@ -0,0 +1,284 @@
+    sorted_indices.resize(bins.numItems());
+    auto* sorted_indices_data = sorted_indices.data();
+
+    const auto Ntheta = m_ntheta;
+    const auto Nphi = m_nphi;
+
+    auto dr = m_delta_ur;
+    auto dtheta = 2.0_prt * MathConst::pi / Ntheta;
+    auto dphi = MathConst::pi / Nphi;
+    auto c2 = PhysConst::c * PhysConst::c;
+
+    auto heapSort = HeapSort();
+
+    // Loop over cells
+    amrex::ParallelForRNG( n_cells,
+        [=] AMREX_GPU_DEVICE (int i_cell, amrex::RandomEngine const& engine) noexcept
+        {
+            // The particles that are in the cell `i_cell` are
+            // given by the `indices[cell_start:cell_stop]`
+            const auto cell_start = static_cast<int>(cell_offsets[i_cell]);
+            const auto cell_stop  = static_cast<int>(cell_offsets[i_cell+1]);
+            const auto cell_numparts = cell_stop - cell_start;
+
+            // do nothing for cells with less particles than min_ppc
+            // (this intentionally includes skipping empty cells, too)
+            if (cell_numparts < min_ppc) {
+                return;
+            }
+
+            // Loop over particles and label them with the appropriate momentum bin
+            // number. Also assign initial ordering to the sorted_indices array.
+            for (int i = cell_start; i < cell_stop; ++i)
+            {
+                // get polar components of the velocity vector
+                auto u_mag = std::sqrt(
+                    ux[indices[i]]*ux[indices[i]] +
+                    uy[indices[i]]*uy[indices[i]] +
+                    uz[indices[i]]*uz[indices[i]]
+                );
+                auto u_theta = std::atan2(uy[indices[i]], ux[indices[i]]) + MathConst::pi;
+                auto u_phi = std::acos(uz[indices[i]]/u_mag);
+
+                const auto ii = static_cast<int>(u_theta / dtheta);
+                const auto jj = static_cast<int>(u_phi / dphi);
+                const auto kk = static_cast<int>(u_mag / dr);
+
+                momentum_bin_number_data[i] = ii + jj * Ntheta + kk * Ntheta * Nphi;
+                sorted_indices_data[i] = i;
+            }
+
+            // sort indices based on comparing values in momentum_bin_number
+            heapSort(sorted_indices_data, momentum_bin_number_data, cell_start, cell_numparts);
+
+            // initialize variables used to hold cluster totals
+            int particles_in_bin = 0;
+            amrex::ParticleReal total_weight = 0._prt, total_energy = 0._prt;
+#if !defined(WARPX_DIM_1D_Z)
+            amrex::ParticleReal cluster_x = 0._prt;
+#endif
+#if defined(WARPX_DIM_3D)
+            amrex::ParticleReal cluster_y = 0._prt;
+#endif
+            amrex::ParticleReal cluster_z = 0._prt;
+            amrex::ParticleReal cluster_ux = 0._prt, cluster_uy = 0._prt, cluster_uz = 0._prt;
+
+            // Finally, loop through the particles in the cell and merge
+            // ones in the same momentum bin
+            for (int i = cell_start; i < cell_stop-1; ++i)
```

Comment:
This loop is skipping the last particle in each cell. It looks like that last particle would never get merged, even if it is in the same velocity bin as the previous particles.
If you switch the checks below, this loop can include the last particle because the `if` check would be short circuited and `sorted_indices_data[i + 1]]` won't be evaluated.
```
if ((i == cell_stop - 1) ||
                    (momentum_bin_number_data[sorted_indices_data[i]] != momentum_bin_number_data[sorted_indices_data[i + 1]]))
```

Reply:
Good point! Thanks for the suggestion.